libgit/autogit_manager.go

```diff
@@ -51,12 +60,108 @@ func NewAutogitManager(config libkbfs.Config) *AutogitManager {
 		registeredFBs:          make(map[libkbfs.FolderBranch]bool),
 		repoNodesForWatchedIDs: make(map[libkbfs.NodeID]*repoDirNode),
+		deleteCancels:          make(map[string]context.CancelFunc),
 	}
 }
 
 // Shutdown shuts down this manager.
 func (am *AutogitManager) Shutdown() {
-	// No-op.
+	am.registryLock.Lock()
+	defer am.registryLock.Unlock()
+	am.shutdown = true
+	for _, cancel := range am.deleteCancels {
+		cancel()
+	}
+}
+
+func (am *AutogitManager) removeOldCheckoutsForHandle(
+	ctx context.Context, h *libkbfs.TlfHandle, branch libkbfs.BranchName) {
+	// Make an "unwrapped" FS, so we don't end up recursively entering
+	// the virtual autogit nodes again.
+	fs, err := libfs.NewUnwrappedFS(
+		ctx, am.config, h, branch, "", "", keybase1.MDPriorityNormal)
+	if err != nil {
+		am.log.CDebugf(ctx, "Error making unwrapped FS for TLF %s: %+v",
+			h.GetCanonicalPath(), err)
+		return
+	}
+
+	fi, err := fs.Stat(AutogitRoot)
+	if os.IsNotExist(errors.Cause(err)) {
+		// No autogit repos to remove.
+		return
+	} else if err != nil {
+		am.log.CDebugf(ctx,
+			"Error checking autogit in unwrapped FS for TLF %s: %+v",
+			h.GetCanonicalPath(), err)
+		return
+	}
+
+	ctx, ok := func() (context.Context, bool) {
+		am.registryLock.Lock()
+		defer am.registryLock.Unlock()
+		if am.shutdown {
+			return nil, false
+		}
+		p := h.GetCanonicalPath()
+		if _, ok := am.deleteCancels[p]; ok {
+			return nil, false
+		}
+
+		ctx, cancel := context.WithCancel(ctx)
+		am.deleteCancels[p] = cancel
+		return ctx, true
+	}()
+	if !ok {
+		return
+	}
+
+	am.log.CDebugf(ctx, "Recursively deleting old autogit data in TLF %s",
+		h.GetCanonicalPath())
+	defer func() {
+		am.log.CDebugf(ctx, "Recursive delete of autogit done: %+v", err)
+		am.registryLock.Lock()
+		defer am.registryLock.Unlock()
+		delete(am.deleteCancels, h.GetCanonicalPath())
+	}()
+	err = recursiveDelete(ctx, fs, fi)
+}
+
+func (am *AutogitManager) removeOldCheckouts(node libkbfs.Node) {
+	ctx := libkbfs.CtxWithRandomIDReplayable(
+		context.Background(), ctxAutogitIDKey, ctxAutogitOpID, am.log)
+
+	h, err := am.config.KBFSOps().GetTLFHandle(ctx, node)
+	if err != nil {
+		am.log.CDebugf(ctx, "Error getting handle: %+v", err)
+		return
+	}
+
+	am.removeOldCheckoutsForHandle(ctx, h, node.GetFolderBranch().Branch)
+}
+
+func (am *AutogitManager) removeSelfCheckouts() {
+	ctx := libkbfs.CtxWithRandomIDReplayable(
+		context.Background(), ctxAutogitIDKey, ctxAutogitOpID, am.log)
+
+	session, err := am.config.KBPKI().GetCurrentSession(ctx)
+	if err != nil {
+		am.log.CDebugf(ctx,
+			"Unable to get session; ignoring self-autogit delete: +%v", err)
+		return
+	}
+
+	h, err := libkbfs.GetHandleFromFolderNameAndType(
+		ctx, am.config.KBPKI(), am.config.MDOps(),
+		string(session.Name), tlf.Private)
+	if err != nil {
+		am.log.CDebugf(ctx,
+			"Unable to get private handle; ignoring self-autogit delete: +%v",
+			err)
+		return
+	}
+
+	am.removeOldCheckoutsForHandle(ctx, h, libkbfs.MasterBranch)
```

Comment:
Do we need to call `takeConfigLock` here like in https://github.com/keybase/kbfs/blob/5cac0cd933f685215f2eca6d9b7d0e0961e2137f/libgit/repo.go#L814 ? Same above.

Reply:
That lock is for the actual `.kbfs_git` repos, and not the autogit checkouts.

You could argue that we need to take some sort of autogit lock, but if there are two new clients both trying to remove all files at the same time, that seems pretty safe to me.  And if there are an old client and a new client racing to access autogit, that also seems fine since we've agreed there are basically zero users of autogit, besides kbpbot.

Does that make sense?

Reply:
That totally makes sense. Thanks!